src/librustc_mir/borrow_check.rs

```diff
@@ -373,10 +373,19 @@ impl<'cx, 'gcx, 'tcx> DataflowResultsConsumer<'cx, 'tcx> for MirBorrowckCtxt<'cx
                                      Consume, (value, span), flow_state);
             }
 
-            TerminatorKind::Goto { target: _ } |
             TerminatorKind::Resume |
             TerminatorKind::Return |
-            TerminatorKind::GeneratorDrop |
+            TerminatorKind::GeneratorDrop => {
+                for local in self.mir.local_decls.indices() {
+                    self.access_lvalue(
+                        ContextKind::StorageDead.new(loc),
+                        (&Lvalue::Local(local), span),
+                        (Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
+                        flow_state
+                    );
```

Comment:
This PR is also aiming to check for in-scope loans of "thread-local statics" and report an error. It may be that it already prevents those. To check, can you add the following test as `src/test/compile-fail/borrowck/thread-local-static-borrow-outlives-fn.rs` (I don't believe we have any such test already):

```rust
// Test that borrows of `#[thread_local]` statics cannot
// outlive the function. See also #17954.

// revisions: ast mir

#![feature(thread_local)]

#[thread_local]
static FOO: u8 = 3;

fn assert_static(_t: &'static u8) {}
fn main() {
     assert_static(&FOO);
}
```

This currently gives a [somewhat obscure error about the lifetime of a temporary value](https://play.rust-lang.org/?gist=a66257a3be6b7fdccdfa8fb257de8062&version=nightly), which makes me think that it will likely generate an error under your PR as well, presuming MIR introduces a temporary at the same spot. I'm actually a bit unsure just *why* we are generating a temporary here -- presumably something to do with how we lower access to `#[thread_local]` statics. Anyway, let's add the test. If it gets an error in both checkers, great, otherwise, we'll have to add a bit of code right around here I think to check for outstanding borrows of `#[thread_local]` statics.